include/boost/geometry/strategies/transform/matrix_transformers.hpp

```diff
@@ -41,6 +44,102 @@ namespace boost { namespace geometry
 namespace strategy { namespace transform
 {
 
+namespace detail { namespace matrix_transformer
+{
+
+template
+<
+    typename CalculationType,
+    typename P2,
+    std::size_t Dimension2,
+    std::size_t Dimension2Count
+>
+struct set_point_from_vec
+{
+    typedef CalculationType ct;
+    static inline void apply(P2& p2, qvm::vec<ct, Dimension2 + 1>& p2temp)
+    {
+        typedef typename geometry::coordinate_type<P2>::type ct2;
+        set<Dimension2Count>(p2, boost::numeric_cast<ct2>(qvm::A<Dimension2Count>(p2temp)));
+        set_point_from_vec<CalculationType, P2, Dimension2, Dimension2Count-1>::apply(p2, p2temp);
+    }
+};
+
+template
+<
+    typename CalculationType,
+    typename P2,
+    std::size_t Dimension2
+>
+struct set_point_from_vec<CalculationType, P2, Dimension2, -1>
+{
+    typedef CalculationType ct;
+    static inline void apply(P2& p2, qvm::vec<ct, Dimension2 + 1>& p2temp) {}
+};
+
+template
+<
+    typename CalculationType,
+    typename P1,
+    std::size_t Dimension1,
+    std::size_t Dimension1Count
+>
+struct set_vec_from_point
+{
+    typedef CalculationType ct;
+    static inline void apply(const P1& p1, qvm::vec<ct, Dimension1 + 1>& p1temp)
+    {
+        qvm::A<Dimension1Count>(p1temp) = get<Dimension1Count>(p1);
+        set_vec_from_point<CalculationType, P1, Dimension1, Dimension1Count - 1>::apply(p1, p1temp);
+    }
+};
+
+template
+<
+    typename CalculationType,
+    typename P1,
+    std::size_t Dimension1
+>
+struct set_vec_from_point<CalculationType, P1, Dimension1, -1>
+{
+    typedef CalculationType ct;
+    static inline void apply(const P1& p1, qvm::vec<ct, Dimension1 + 1>& p1temp) {}
+};
+
+
+template
+<
+    typename CalculationType,
+    std::size_t Dimension1,
+    std::size_t Dimension2
+>
+class matrix_transformer
+{
+protected :
+    typedef CalculationType ct;
+    typedef boost::qvm::mat<ct, Dimension2 + 1, Dimension1 + 1> matrix_type;
+    matrix_type m_matrix;
+public :
+
+    matrix_type const& matrix() const { return m_matrix; }
+    template <typename P1, typename P2>
+    inline bool apply(P1 const& p1, P2& p2) const
+    {
+        assert_dimension<P1,Dimension1>();
+        assert_dimension<P2,Dimension2>();
+        qvm::vec<ct,Dimension1 + 1> p1temp;
+        qvm::vec<ct,Dimension2 + 1> p2temp;
+        set_vec_from_point<ct, P1, Dimension1, Dimension1 - 1>::apply(p1, p1temp);
+        qvm::A<Dimension1>(p1temp) = 1;
+        p2temp = m_matrix * p1temp;
```

Comment:
Previously all of this code was "hardcoded" and then simply set e.g.:

    ct p2x = c1 * qvm::A<0,0>(m_matrix) + c2 * qvm::A<0,1>(m_matrix) + qvm::A<0,2>(m_matrix);

Have you checked if there is a difference in performance?

Another thing is that maybe it would be possible to write a wrapper or a view for Boost.Geometry point, (holding a reference or pointer to Point) and adapter to QVM Vector concept. This could allow you to pass it into QVM without conversions. So maybe all of the code above could look like:

    qvm_vector_view<P2> v2(p2);
    v2 = m_matrix * qvm_vector_view<P1 const>(p1);

Reply:
Here is an example of such view type:

https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/views/detail/indexed_point_view.hpp#L35

This is a view for segments and boxes which is adapted to Point concept and allows to represent one of segment's endpoints or box corner.

Reply:
I did consider such an approach, but I was unsure how to exactly fit the boost::numeric_cast into that, because qvm seems to require write access through a returned reference, rather than passing the value to be written to some user-defined function where I could do the numeric_cast. Because I didn't want to change behavior if possible and I was concerned I could overlook some effect of numeric_cast, I did it this way and first calculated all entries, then wrote them to the result vector with the numeric_cast.

Reply:
> by doing the full multiplication

The easiest solution is to keep the hardcoded operations as they were implemented in specializations and use your new implementation for other/higher dimensions. This actually would solve both problems (unneeded operations and temporary vectors)

>Regarding a vector view

Right, I didn't check this well enough, QVM returns mutable reference and Geometry uses set function. So indeed representing Point as Vector may be not possible. Theoretically it would be possible to return non-true reference (like `boost::refference_wrapper`) but I wouldn't be so sure that it would work with QVM. I'd rather assume it requires true references. So yes, temporaries may be the best solution.

Reply:
Ok. In my latest commit I've reintroduced the hardcoded operations with the following changes:
- the 3d to 2d transformations now considers the z coordinate in the input as discussed on the matter of matrix dimensions for 3d to 2d.
- I removed some temporaries in 2d to 2d that were not included in the original implementation of 3d to 3d.

I previously didn't notice, that only dimension_greater_equal is asserted in https://github.com/boostorg/geometry/blob/5a13eea5edbad79c9512dd66ac18e6c0b553b489/include/boost/geometry/strategies/transform/matrix_transformers.hpp#L95 . I suppose this is meant to support the case that, for example, a 2d to 2d transformation is applied to 3d points that leaves the z coordinate untouched? In case this is intended, I changed the asserts into only asserting greater or equal.